### DISCUSSION TITLE
Welcome to vscode-masm-runner Discussions!

### DISCUSSION THREAD

**Discussion by istareatscreens** [istareatscreens/vscode-masm-runner#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by stardustgd** [comment]
Hello, I've been using this extension while I am learning x86 assembly and it has been super helpful! I was wondering how can I set up this extension to use several asm files in the same workspace.

**Reply by istareatscreens** [reply]
Hello, I am glad you are enjoying it.  It should be possible although I have never done such a thing with it.  Are you using native windows compilation or the wine emulator? I have never actually had to link .asm files and it has been awhile since I have actually done anything in MASM so based on a short google this seems to be how it is done:
https://stackoverflow.com/questions/61674766/linking-multiple-files-with-masm-and-visual-studio-2019
If this is not the way you are attempting/correct please share the textbook method of how you are attempting to achieve it. Given you don't need to modify the link/compilation command flags it should work as long as you use absolute paths.

**If you are using the wine emulator:**
- Currently when you send files to the emulator they are located at the path:
D:/
So make sure all files that are to be linked are added to the wine emulator filesystem (this could be automated by reading the file in the extension by reading .asm files for INCLUDE statements but this is currently not implemented).
INCLUDE D:/fileToLink.inc

Basically I send messages from the extension to the webview (compiling command is triggered here):
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/src/extension.ts#L112
and then convert the messages to specific events which runs the functions that actually interact directly with the wine emulator (webview via wasm) located here:
https://github.com/istareatscreens/vscode-masm-runner/blob/master/src/js/components/boxedwine/Boxedwine.jsx

**If you are compiling natively** 
Your include statement simply should be the absolute path of your project directory. If you open terminal and type pwd (in powershell) you can get the absolute path to the project directory from there simply appending the file name to it in the include statement should work (This too could be potentially automated with some simple regex reading the include statement)

When compiling from native windows this is the code that is run (Some regex could be used to find and replace the path directory automatically):
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/src/extension.ts#L166

If you require modifications of the command line flags/linking flags to actually compile things I could add a setting for that pretty easily, although this wouldn't be an ideal setup it would allow you to test things and if you got it working I could probably write a more convenient solution by reading the source files and updating the compilation/linking command dynamically.
Documentation on compiling and linking commands can be found here although some googling might help:
https://www.japheth.de/JWasm.html
https://www.japheth.de/JWasm/Manual.html
https://www.japheth.de/JWlink.html
https://www.japheth.de/JWlink/JWlink.htm

If you want me to add the ability to add compiler flags dynamically via settings it shouldn't take me long. 

**tl:dr** - The current version of this may not support linking .asm files (I have never attempted this/not really sure how to do this) if it requires modification to compiler and linking flags. 

If by chance it does not require modification of the compiling/linking commands then it should work by just using the absolute path to your workspace (run pwd (also make sure there are no spaces in the directory path they usually cause issues) from vscode terminal in powershell) OR if using the webview D:/ with your INCLUDE statement and make sure the files are sent to the webview. 

If it requires compiler/linking command line argument modification I would gladly implement a setting to allow modification of them so you could provide me with a working use case so I could implement this use case OR if you are interested in implementing the functionality yourself I would provide you support on how to implement it.

**Reply by stardustgd** [reply]
Hello, thank you for your response. I am compiling natively on Windows. The Irvine textbook's method of the multimodule program uses a .inc file that defines all the prototypes and several different .asm files that implement those prototypes. Currently, I am testing this using Irvine's example from chapter 8 [ModSum32_advanced](https://github.com/istareatscreens/vscode-masm-runner/files/9998510/ModSum32_advanced.zip). I have tried using absolute paths that don't have spaces, but I get the following error during linking:
```
JWlink Version 1.9beta 13
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
loading object files
searching libraries
Error! E2028: _PromptForIntegers@12 is an undefined reference
Error! E2028: _ArraySum@8 is an undefined reference
Error! E2028: _DisplaySum@8 is an undefined reference
creating a PE executable
file c:\Users\\Sum_main.temp.obj(c:\Users\\Sum_main.temp.asm): undefined symbol _PromptForIntegers@12
file c:\Users\\Sum_main.temp.obj(c:\Users\\Sum_main.temp.asm): undefined symbol _ArraySum@8        
file c:\Users\\Sum_main.temp.obj(c:\Users\\Sum_main.temp.asm): undefined symbol _DisplaySum@8
```
I have come up with a solution: to run JWASM for every .asm file in the workspace, then use JWlink on all the .obj files to produce a single .exe file, so I think it might require some modification to the linking command line arguments. I would greatly appreciate it if you could  implement a setting for this (if possible)!

**Reply by istareatscreens** [reply]
Hey I will throw together a dirty fix for this so you can apply compiler flags to the JWASM and JWLINK commands through settings as a temporary solution for you to test/get it working, I will try to get this done tonight as it should be fairly trivial. If you get it to work please let me know exactly what you entered. I will then implement a better solution this week or next by likely reading the .inc files and adding the flags dynamically.

I will let you know when I push the changes :)

**Reply by istareatscreens** [reply]
Alright I published version 0.4.5 the settings I added are currently undocumented and only added for native compiling/linking

![image](https://user-images.githubusercontent.com/46049787/201593256-1f1239d9-e38b-4ff0-aba5-1f0b2938485b.png)

Here they are you should be able to link things properly try not to add spaces to your file names or directories as I did not apply the directory fixing for that.

I hope you figure it out, let me know when you do and I will implement a better solution

Some Notes:
These are the settings added:
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/package.json#L101

They are loaded from settings into the extension here:
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/src/extension.ts#L196

Compile command: 
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/src/extension.ts#L273

Linking command:
https://github.com/istareatscreens/vscode-masm-runner/blob/master/extension/src/extension.ts#L276

This ^ should show you where each setting is inserted into

**Reply by stardustgd** [reply]
I think I am able to get it working. I added the additional .asm files other than the main asm file to the compiler arguments at the end. I did the same for the .obj files to the link arguments at the end. It seems like this method will vary depending on how many files there are in the workspace. 

![image](https://user-images.githubusercontent.com/88991140/201606447-8918a6b7-354f-44c1-bd54-f0ace49e8afd.png)

I also have to ensure that every file that has the `INCLUDE Irvine32.inc` line is changed to the absolute path of 
`INCLUDE C:\Users\.vscode\extensions\istareatscreens.masm-runner-0.4.3\native\irvine\Irvine32.inc` (In the Irvine example, the include statement is in the sum.inc file, rather than the main asm file). 

![image](https://user-images.githubusercontent.com/88991140/201604636-ac835e81-f2c0-4a3f-8346-515471e1cf76.png)

But it does seem to be assembling and linking all the files in the workspace properly after these arguments are added.

![image](https://user-images.githubusercontent.com/88991140/201604707-d650c17b-4623-4a38-9a36-084a1d0bcbbb.png)

I also discovered that you can assemble and link all files in a workspace through `JWASM.EXE  /Zd /coff *.asm` to assemble and `JWlink.exe ... file *.obj` to link. Maybe this way could allow for a universal fix, although this may cause conflicts with the .temp files that are created when the extension is run. 

I really appreciate your responses and help with this thus far!

**Reply by istareatscreens** [reply]
I am glad you were able to get it working and I am happy to have helped! 

`
I also have to ensure that every file that has the INCLUDE Irvine32.inc line is changed to the absolute path of
INCLUDE C:\Users\.vscode\extensions\istareatscreens.masm-runner-0.4.3\native\irvine\Irvine32.inc (In the Irvine example, the include statement is in the sum.inc file, rather than the main asm file).
`
This is expected as I am reading the current file and changing the INCLUDE Irvine32.inc to the absolute path. This extension was written pretty hastily I definitely need to refine this further.

Is the .inc file required to compile and link properly? 

If you could share your code with me for testing I would greatly appreciate it, I of course would not share it with anyone else. If that is not possible I completely understand. 

Thanks you I really appreciate you bringing this issue to my attention  :) and wish you all the best in your studies.

**Reply by stardustgd** [reply]
I believe the .inc file is needed for the procedure declarations, constants/structs declarations, and the inclusion of the Irvine library. As of right now, I'm using [ModSum32_advanced](https://github.com/istareatscreens/vscode-masm-runner/files/10001009/ModSum32_advanced.zip) from Irvine's library examples which has all the .asm files and a .inc file. Feel free to test with this example!
```
_arrysum.asm (procedure implementation)
_display.asm (procedure implementation)
_prompt.asm (procedure implementation)
sum.inc (procedure declarations)
Sum_main.asm (main file)
```
